### DISCUSSION TITLE
Stochastic Oscillator

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#237] [Indicators]
Community discussion about [Stochastic Oscillator](https://dotnet.stockindicators.dev/indicators/Stoch/#content) and KDJ Index

**Comment by DaveSkender** [comment]
I've just added the KDJ Index %J value to Stochastic Oscillator.  See #315 for the background and assumptions.

**Comment by vebollu** [comment]
![image](https://user-images.githubusercontent.com/77417916/105112122-f1475200-5a87-11eb-877d-ef526b5f2d27.png)

**Comment by vebollu** [comment]
Can you please add Fast Stochastic property too?

**Reply by DaveSkender** [reply]
A fast variant can be done with parameter argument `smoothingPeriod=1`

**Reply by vebollu** [reply]
ty, it got worked.

**Comment by netprok** [comment]
Hi, thank you for your efforts and help. 

I am using your np and trying to get KDJ results, but the PercentJ and Signal are missing. 



![image](https://user-images.githubusercontent.com/31617979/125477829-60013fbe-eea1-4836-8b16-6e01aaff3e7b.png)

![image](https://user-images.githubusercontent.com/31617979/125478127-1587078e-0c90-43c9-9324-95c939df5d5b.png)

**Reply by DaveSkender** [reply]
Most indicators have a warmup period where results cannot be calculated on the front end of your historical quotes, so they are null.  With the parameters shown, I'd expect the first 18-19 periods to have `null` %J results.

If you want the library to automatically exclude the recommended amount of warmup periods, use the `.RemoveWarmupPeriods()` extension:

```csharp
// usage if you want to remove the 
// recommended warmup periods
IEnumerable<StochResult> results =  
  quotes.GetStoch(14, 3, 3)
    .RemoveWarmupPeriods(); 
```

See the [Utilities Guide](https://dotnet.stockindicators.dev/utilities#remove-warmup-periods) for more information.

**Reply by DaveSkender** [reply]
```csharp
r.PercentJ = (3 * r.Oscillator) - (2 * r.Signal);
```
Given how %J uses oscillator and signal values, I suspect you're not using enough historical quotes.  The bare minimum you'd need to get all three values is `lookbackPeriods + smoothPeriods + signalPeriods`, so you'd have to update your `NeededCandlesCount` if you are using that to limit the size of your quotes.

With that said, I think most people will use larger datasets in general, not just the bare minimum.

**Comment by DaveSkender** [comment]
Another discussion on KDJ, compared to other charting platforms - see #612.

If anyone can find source materials from the original author, [George Lane](https://en.wikipedia.org/wiki/George_Lane_(technical_analyst)), let me know.  It would help resolve the debate over which recipe to use.  

There are a few opinionated variants out there and they all use a slightly different approach; none of them being more "accurate" than another, just slightly different.
